### ISSUE TITLE
IOU Tensor Representation

### ISSUE THREAD

**Issue by ShafinH** [ultralytics/yolov5#8603]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I want to update the loss function based on IOU values and was wondering what do the IOU tensors in the loss calculation represent.
```Python
pxy = pxy.sigmoid() * 2 - 0.5
pwh = (pwh.sigmoid() * 2) ** 2 * anchors[i]
pbox = torch.cat((pxy, pwh), 1)  # predicted box
iou = bbox_iou(pbox, tbox[i], CIoU=True).squeeze()  # iou(prediction, target) 
print(iou.size())
print(iou)
lbox += (1.0 - iou).mean()  # iou loss
```
Added the `print(iou)` and for my custom dataset it outputs tensors of large sizes like this for example
```
torch.Size([30])
tensor([0.24051, 0.27898, 0.23945, 0.13778, 0.42550, 0.19172, 0.40641, 0.20556, 0.25486, 0.19675, 0.22835, 0.15716, 0.10606, 0.34960, 0.03668, 0.06333, 0.12623, 0.17952, 0.17402, 0.09726, 0.27948, 0.12250, 0.24838, 0.14936, 0.11186, 0.03941, 0.00368, 0.03528, 0.18717, 0.15006], device='cuda:0', grad_fn=<SqueezeBackward0>) 
```

So what does this tensor represent? Is it the IOU of the predicted box with all the ground truth labels of the given frame?

**Comment by glenn-jocher** [comment]
@ShafinH yes this is the predicted-true box CIoU tensor

**Comment by sezer-muhammed** [comment]
Hi, Did you manage it? how and where did you change to change metric IOU? I also need the same thing.

**Comment by ShafinH** [comment]
> @ShafinH yes this is the predicted-true box CIoU tensor

@glenn-jocher  Hi, thanks for the response, but I had a few more questions.

1) Why are the tensors so big, because there shouldn't be this many target boxes per frame, so is this the total amount of boxes within 1 batch?

2) Where can I find the IOU for the hundreds/thousands of anchor box proposals? Is there another place where the loss is updated based on the IOU of the anchor box proposals with the target boxes, or is it only updated in the code above?

3) Are these the final predicted boxes after NMS happens, and is the loss only updated on these final boxes?

Thanks

**Comment by ShafinH** [comment]
@glenn-jocher Could you help with those questions?

**Comment by glenn-jocher** [comment]
@ShafinH training never uses NMS, this is only for inference. The size of the tensor is the size of the tensor, what other size would you want it to be?

**Comment by ShafinH** [comment]
I meant what exactly is in each specific tensor? Is it the IOU for every predicted box within a single batch, or a single image?

**Comment by glenn-jocher** [comment]
@ShafinH all images in a batch are processed together.
